### PR TITLE
Feature/wa 284 architecture refactoring

### PR DESCRIPTION
**Pull request by DmitryBespalov**
Introduced `IdentityApplicationService`

### CODE REVIEW

File: TestTools/TestSupport.swift [hunk 1]
```diff
@@ -30,20 +30,20 @@ final class TestSupport {
                 resettableObjects.forEach { $0.resetAll() }
             case ApplicationArguments.setPassword:
                 if let password = iterator.next() {
-                    try? account.setMasterPassword(password)
+                    try? identityService.registerUser(RegisterUserCommand(password){})
                 }
             case ApplicationArguments.setSessionDuration:
                 if let duration = timeInterval(&iterator) {
-                    account.sessionDuration = duration
+                    identityService.configureSession(duration)
```
Comment (sche) [inline TestTools/TestSupport.swift]: why session should be part of identity?

File: TestTools/TestSupport.swift [hunk 2]
```diff
@@ -30,20 +30,20 @@ final class TestSupport {
                 resettableObjects.forEach { $0.resetAll() }
             case ApplicationArguments.setPassword:
                 if let password = iterator.next() {
-                    try? account.setMasterPassword(password)
+                    try? identityService.registerUser(RegisterUserCommand(password){})
```
Comment (sche) [inline TestTools/TestSupport.swift]: It looks scary

File: safe/safe/AppFlow/AppFlowCoordinator.swift [hunk 3]
```diff
@@ -13,25 +13,25 @@ protocol AppFlowCoordinatorProtocol: class {
...
     init(account: AccountProtocol = Account.shared) {
-        self.account = account
+        identityService = IdentityApplicationService(account: account)
         onboardingFlowCoordinator = OnboardingFlowCoordinator(account: account)
     }
 
     func startViewController() -> UIViewController {
         lockedViewController = onboardingFlowCoordinator.startViewController()
-        if account.hasMasterPassword {
+        if identityService.hasPrimaryUser() {
```
Comment (sche) [inline safe/safe/AppFlow/AppFlowCoordinator.swift]: Confusing for me. We instantiate identity service with account. What is a primary user in this case?

File: safe/safe/AppFlow/AppFlowCoordinator.swift [hunk 4]
```diff
@@ -40,7 +40,7 @@ final class AppFlowCoordinator: AppFlowCoordinatorProtocol {
     }
 
     func unlockController(completion: @escaping () -> Void) -> UIViewController {
-        return UnlockViewController.create(account: account, completion: completion)
+        return UnlockViewController.create(account: identityService.account, completion: completion)
```
Comment (sche) [inline safe/safe/AppFlow/AppFlowCoordinator.swift]: why not passing identityService?

File: TestTools/TestSupport.swift (hunk 1 above)
Comment (DmitryBespalov) [reply TestTools/TestSupport.swift]: Ideally, these settings are related to the Configuration concept. I'll refactor it in the next PR.

File: TestTools/TestSupport.swift (hunk 2 above)
Comment (DmitryBespalov) [reply TestTools/TestSupport.swift]: I'd like to use command pattern for any changes to the model. Would this code look better?:

    try? identityService.registerUser(command: RegisterUserCommand(with: password))

File: safe/safe/AppFlow/AppFlowCoordinator.swift (hunk 3 above)
Comment (DmitryBespalov) [reply safe/safe/AppFlow/AppFlowCoordinator.swift]: True for now, but I'll refactor it back to the `User` when we'll have User domain concept.

File: safe/safe/AppFlow/AppFlowCoordinator.swift (hunk 4 above)
Comment (DmitryBespalov) [reply safe/safe/AppFlow/AppFlowCoordinator.swift]: This is the next step - for the next PR.

File: safe/safe/AppFlow/AppFlowCoordinator.swift (hunk 3 above)
Comment (DmitryBespalov) [reply safe/safe/AppFlow/AppFlowCoordinator.swift]: But wait... there is already `registerUser`, `authenticateUser`.. maybe his should be `hasRegisteredUser`?

File: safe/safe/ApplicationLayer/IdentityApplicationService.swift [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    var completion: RegisterCompletion
+    typealias RegisterCompletion = () -> Void
+
+    init(_ password: String, completion: @escaping RegisterCompletion) {
+        self.password = password
+        self.completion = completion
+    }
+}
+
+class IdentityApplicationService {
```
Comment (sche) [inline safe/safe/ApplicationLayer/IdentityApplicationService.swift]: final

Comment (DmitryBespalov) [reply safe/safe/ApplicationLayer/IdentityApplicationService.swift]: no, this should be subclassable for testing purposes

**Review by sche — APPROVED** [review]

**Pull request merged**